File: examples/canonical-gcp-functions/eventarc-monitoring-router/go/src/handler.go
package handler

import (
	"context"
	"encoding/json"
	"fmt"
	"time"
)

type IncidentEvent struct {
	ID      string `json:"id"`
	Service string `json:"service"`
	State   string `json:"state"`
	Summary string `json:"summary"`
}

type Clients interface {
	Claim(ctx context.Context, key string) (bool, error)
	LookupOwner(ctx context.Context, service string) (string, error)
	WriteAudit(ctx context.Context, row AuditRow) error
	SendSlack(ctx context.Context, owner string, event IncidentEvent) error
}

type AuditRow struct {
	IncidentID string
	Service    string
	Owner      string
	State      string
	At         time.Time
}

type Result struct {
	Routed    bool
	Duplicate bool
	Owner     string
}

func Handle(ctx context.Context, event IncidentEvent, clients Clients) (Result, error) {
	started := time.Now()
	dedupeKey := event.ID + ":" + event.State
	claimed, err := clients.Claim(ctx, dedupeKey)
	if err != nil {
		return Result{}, err
	}
	if !claimed {
		logShape(started, event.ID, dedupeKey, "DROP", "duplicate incident suppressed")
		return Result{Duplicate: true}, nil
	}
	owner, err := clients.LookupOwner(ctx, event.Service)
	if err != nil {
		return Result{}, err
	}
	row := AuditRow{IncidentID: event.ID, Service: event.Service, Owner: owner, State: event.State, At: time.Now().UTC()}
	if err := clients.WriteAudit(ctx, row); err != nil {
		return Result{}, err
	}
	if err := clients.SendSlack(ctx, owner, event); err != nil {
		return Result{}, err
	}
	logShape(started, event.ID, dedupeKey, "ALLOW", "incident routed")
	return Result{Routed: true, Owner: owner}, nil
}

func logShape(started time.Time, correlationID, dedupeKey, decision, msg string) {
	entry := map[string]any{
		"timestamp":     time.Now().UTC().Format(time.RFC3339),
		"severity":      "INFO",
		"msg":           msg,
		"trace_id":      "local-trace",
		"request_id":    correlationID,
		"correlation_id": correlationID,
		"provider":      "gcp",
		"runtime_tier":  "function",
		"function_name": "accb-dev-gcp-monitoring-router",
		"dedupe_key":    dedupeKey,
		"decision":      decision,
		"latency_ms":    time.Since(started).Milliseconds(),
	}
	payload, _ := json.Marshal(entry)
	fmt.Println(string(payload))
}

File: examples/canonical-gcp-functions/pubsub-translation-stream/python/src/handler.py
from __future__ import annotations

import json, os, time
from dataclasses import dataclass
from typing import Any, Protocol

class FirestoreClient(Protocol):
    def claim_message(self, dedupe_key: str, payload: dict[str, Any]) -> bool: ...
    def write_translation(self, dedupe_key: str, payload: dict[str, Any]) -> None: ...
class TranslateClient(Protocol):
    def translate(self, text: str, target_locale: str) -> str: ...
class PubSubClient(Protocol):
    def publish(self, topic: str, payload: dict[str, Any]) -> None: ...
@dataclass
class Clients:
    firestore: FirestoreClient
    translate: TranslateClient
    pubsub: PubSubClient

def handle_pubsub_message(event: dict[str, Any], clients: Clients | None = None) -> dict[str, Any]:
    started = time.time()
    data = event.get("data", {})
    message_id = data["messageId"]
    target_locale = data.get("targetLocale", "es")
    text = data["text"]
    event_id = event.get("id", message_id)
    dedupe_key = f"{message_id}:{target_locale}"
    clients = clients or _gcp_clients()
    if not clients.firestore.claim_message(dedupe_key, data):
        _log(started, event_id, dedupe_key, "DROP", "duplicate pubsub translation")
        return {"ok": True, "duplicate": True, "decision": "DROP"}
    translated = clients.translate.translate(text, target_locale)
    payload = {"message_id": message_id, "target_locale": target_locale, "translated_text": translated, "status": "TRANSLATED"}
    clients.firestore.write_translation(dedupe_key, payload)
    clients.pubsub.publish(os.getenv("COMPLETION_TOPIC", "accb-dev-gcp-translation-complete"), payload)
    _log(started, event_id, dedupe_key, "ALLOW", "translation completed")
    return {"ok": True, "duplicate": False, "decision": "ALLOW", "translated_text": translated}

def _gcp_clients() -> Clients:
    raise RuntimeError("GCP clients are bound by deployment; tests inject clients.")

def _log(started: float, event_id: str, dedupe_key: str, decision: str, msg: str) -> None:
    print(json.dumps({"timestamp": time.strftime("%Y-%m-%dT%H:%M:%SZ", time.gmtime()), "severity": "INFO", "msg": msg, "trace_id": os.getenv("TRACEPARENT", "local-trace"), "request_id": event_id, "correlation_id": event_id, "provider": "gcp", "runtime_tier": "function", "function_name": os.getenv("FUNCTION_TARGET", "accb-dev-gcp-pubsub-translation-stream"), "dedupe_key": dedupe_key, "decision": decision, "latency_ms": int((time.time() - started) * 1000)}))
